### ISSUE TITLE
custom training with arbitrary shaped graph, GraphAttention layes gradient becomes zeros.

### ISSUE THREAD

**Issue by ssbahade** [danielegrattarola/spektral#81]
While using **GraphAttention** layer in the custom training, grad values is zero always.

**Comment by danielegrattarola** [comment]
Hi,
Can you provide a code example to reproduce the issue?
Thanks.

**Comment by ssbahade** [comment]
Hi, Here is the sample code example:

data is having a batch size 1 and each sample is different size as shown below:

A = A_train[start:end][0]      # Binary adjacy matrix size N x N     i.e (6,6), (19,19)...etc
X = x_train[start:end][0]      #  feature matrix                                 i.e (6,3), (19,3)....etc
y = y_train[start:end]          # labels                                               i.e      0,   1 .... etc

       def __init__(self):
                super(GraphNet, self).__init__()
                self.G1 = GraphAttention(32, activation='relu', kernel_regularizer=l2(l2_reg))
                self.G2 = GraphAttention(32, activation='relu', kernel_regularizer=l2(l2_reg))
                self.B1 = tf.keras.layers.BatchNormalization(trainable=True)
                self.A1 = tf.keras.layers.Activation('relu')
                self.pool = GlobalAttentionPool(128)
                self.O2 = tf.keras.layers.Dense(n_classes, activation='softmax', kernel_regularizer=reg, name='dense_2')
         
        def call(self, input_tensor, training=True):
                x = self.G1([input_tensor[0], input_tensor[1]])
                x = self.G2([x, input_tensor[1]])
                x = self.pool(x)
                x = self.O2(x)
                return x

model = GraphNet()
opt = tf.keras.optimizers.Adam(lr=1e-3)
loss_object = tf.keras.losses.SparseCategoricalCrossentropy(from_logits=True)

         with tf.GradientTape(watch_accessed_variables=True) as tape:
                X = (X / 255).astype(np.float32)
                y_pred = model([X, A], training=True)
                y_true = y[0]
                
                loss = loss_object(y_true, y_pred)
         variables = self.trainable_variables
         grads = tape.gradient(loss, variables)
         opt.apply_gradients(zip(grads, variables))

**Comment by danielegrattarola** [comment]
Thanks, I will look at this more in depth. 
Just a heads-up: it seems that you're instantiating the loss with `from_logits=True` but then you have a softmax activation on the last layer. Can this be part of the issue? 

Cheers

**Comment by ssbahade** [comment]
Yes, that is the part of the issue. Thanks for letting me know. It's working now. Is this code operator working for batch size greater than 1.

Sent from Outlook Mobile<https://aka.ms/blhgte>

________________________________

**Comment by danielegrattarola** [comment]
Sure, all layers used in your code support batches!

Cheers
